phantazm-mob/src/main/java/com/github/phantazmnetwork/mob/config/MobModelConfigProcessor.java

```diff
@@ -0,0 +1,131 @@
+
+    private final ConfigProcessor<MinestomDescriptor> descriptorProcessor;
+
+    private final ConfigProcessor<GoalCreator> goalCreatorConfigProcessor;
+
+    private final ConfigProcessor<ItemStack> itemStackConfigProcessor;
+
+    private final MiniMessage miniMessage;
+
+    static {
+        for (EquipmentSlot equipmentSlot : EquipmentSlot.values()) {
+            EQUIPMENT_SLOT_MAP.put(equipmentSlot.name().toUpperCase(), equipmentSlot);
+        }
+    }
+
+    public MobModelConfigProcessor(@NotNull ConfigProcessor<MinestomDescriptor> descriptorProcessor,
+                                   @NotNull ConfigProcessor<GoalCreator> goalCreatorConfigProcessor,
+                                   @NotNull ConfigProcessor<ItemStack> itemStackConfigProcessor,
+                                   @NotNull MiniMessage miniMessage) {
+        this.descriptorProcessor = Objects.requireNonNull(descriptorProcessor, "descriptorProcessor");
+        this.goalCreatorConfigProcessor = Objects.requireNonNull(goalCreatorConfigProcessor,
+                "goalCreatorConfigProcessor");
+        this.itemStackConfigProcessor = Objects.requireNonNull(itemStackConfigProcessor,
+                "itemStackConfigProcessor");
+        this.miniMessage = Objects.requireNonNull(miniMessage, "miniMessage");
+    }
+
+    @Override
+    public @NotNull MobModel dataFromElement(@NotNull ConfigElement element) throws ConfigProcessException {
+        MinestomDescriptor descriptor = descriptorProcessor.dataFromElement(element.getElement("descriptor"));
+
+        ConfigList goalCreatorsGroupsList = element.getListOrThrow("goalCreators");
+        Collection<Iterable<GoalCreator>> goalCreatorsGroups = new ArrayList<>(goalCreatorsGroupsList.size());
+        for (ConfigElement goalCreatorsElement : goalCreatorsGroupsList) {
+            if (!goalCreatorsElement.isList()) {
+                throw new ConfigProcessException("goal creators are not a list");
+            }
+
+            ConfigList goalCreatorsList = goalCreatorsElement.asList();
+            Collection<GoalCreator> goalCreators = new ArrayList<>(goalCreatorsList.size());
+            for (ConfigElement goalCreatorElement : goalCreatorsList) {
+                goalCreators.add(goalCreatorConfigProcessor.dataFromElement(goalCreatorElement));
+            }
+
+            goalCreatorsGroups.add(goalCreators);
+        }
+
+        String displayNameString = element.getStringOrDefault((String) null, "displayName");
+        Component displayName = displayNameString == null ? null : miniMessage.deserialize(displayNameString);
+
+        ConfigNode equipmentNode = element.getNodeOrThrow("equipment");
+        Map<EquipmentSlot, ItemStack> equipment = new HashMap<>(equipmentNode.size());
+        for (Map.Entry<String, ConfigElement> entry : equipmentNode.entrySet()) {
+            EquipmentSlot equipmentSlot = EQUIPMENT_SLOT_MAP.get(entry.getKey().toUpperCase());
```

Comment:
for enums, i have made a general-purpose utility method for constructing cfg processors to serialize/deserialize them, which is also in mapdata
might move that functionality to ethylene

Reply:
that's a bit weird but it would be fine ig